crates/cargo-bsp/tests/integration_test.rs

```diff
@@ -76,3 +81,201 @@ fn exit_notif_without_shutdown() {
+        package_id: &str,
+        features_to_disable: &[&str],
+    ) {
+        let req = bsp_server::Request {
+            id: TEST_REQUEST_ID.into(),
+            method: DisableCargoFeatures::METHOD.into(),
+            params: serde_json::to_value(DisableCargoFeaturesParams {
+                package_id: package_id.to_string(),
+                features: features_to_disable.iter().map(|&s| s.into()).collect(),
+            })
+            .unwrap(),
+        };
+        cl.send(&to_string(&req).unwrap());
+    }
+
+    fn overwrite_cargo_toml_with_features(features_slice: &[(&str, &[&str])]) {
+        const TEST_PROJECT_AUTHOR: &str = "Test Author";
+        const TEST_PROJECT_VERSION: &str = "0.0.1";
+
+        let mut cargo_toml = CargoToml::builder();
+        cargo_toml
+            .name(TEST_PROJECT_NAME)
+            .version(TEST_PROJECT_VERSION)
+            .author(TEST_PROJECT_AUTHOR);
+
+        for &feature_with_deps in features_slice {
+            let mut f = TomlFeature::new(feature_with_deps.0);
+            for &dep in feature_with_deps.1 {
+                f.feature(dep);
+            }
+            cargo_toml.feature(f.build());
+        }
+
+        let cargo_toml_string = cargo_toml.build().unwrap().to_string();
+        File::create("Cargo.toml")
+            .unwrap()
+            .write_all(cargo_toml_string.as_bytes())
+            .unwrap();
+    }
+
+    // Function that creates new temporary project with Cargo.toml with features
+    // and sets it as current directory. The newly created directory has to be deleted.
+    fn create_mock_rust_project_with_features_and_set_it_as_current_dir(
+        features_slice: &[(&str, &[&str])],
+    ) {
+        Command::new(toolchain::cargo())
+            .args(["init", TEST_PROJECT_NAME])
+            .output()
+            .expect("Failed to create new temporary project for testing.");
+
+        std::env::set_current_dir(TEST_PROJECT_NAME).unwrap();
+
+        overwrite_cargo_toml_with_features(features_slice);
+    }
+
+    fn run_test(features_slice: &[(&str, &[&str])], test: fn(&mut Client)) {
+        create_mock_rust_project_with_features_and_set_it_as_current_dir(features_slice);
+
+        spawn_server_with_proper_life_time(test);
+
+        // Do the cleaning
+        std::env::set_current_dir("..").unwrap();
+        remove_dir_all(TEST_PROJECT_NAME).unwrap();
+    }
+
+    fn package_from_response(resp: &str) -> PackageFeatures {
+        let resp: Response = serde_json::from_str(resp).unwrap();
+        let mut current_state: CargoFeaturesStateResult =
+            serde_json::from_value(resp.result.unwrap()).unwrap();
+        assert_eq!(current_state.packages_features.len(), 1);
+        current_state.packages_features.pop().unwrap()
+    }
+
+    fn package_id_from_response(resp: &str) -> String {
+        package_from_response(resp).package_id
+    }
+
+    // Returns pair of available and enabled features from CargoFeatureState response
```

Comment:
to lepiej strukturę na to zrobić i nazwać co to jest zamiast zwracać parę i musieć opisywać w komentarzu